### PR TITLE
bench: add transfer workload

### PR DESCRIPTION
**Pull request by gabriele-0201**
example of usage:
```shell
cd xtask
cargo run -- bench -b nomt,sov-db -n transfer -s 1000 -i 10
```

### CODE REVIEW

**Comment by gabriele-0201** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/109?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#109** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/109?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#108** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/108?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#107** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/107?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#86** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/86?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @gabriele-0201 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: xtask/src/transfer_workload.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    additional_initial_capacity: u64,
+    timer: &mut Timer,
+) {
+    let accounts = (0..n_accounts * 2)
+        .into_iter()
+        .map(|a| a.to_be_bytes().to_vec());
+
+    // prepare additional random entries
+    // they will be hashed so its ok to use just integers
+    let additiona_keys = (n_accounts..n_accounts + additional_initial_capacity)
```
Comment (rphmeier) [inline xtask/src/transfer_workload.rs:18]: typo

File: xtask/src/transfer_workload.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
+use crate::{
+    backend::{Action, DB},
+    timer::Timer,
+};
+
+pub fn transfer_workload(
```
Comment (rphmeier) [inline xtask/src/transfer_workload.rs:6]: Let's adjust this slightly.

There are broadly three things we care about:
  1. The size of the account database
  2. The amount of transfers to perform
  3. The proportion of the database to use for transfers - some state will be warm, some will be cold

You can come up with sub-categories for each of these, but I think we should just change the function for now to accept a parameter for each one of these.

File: xtask/src/transfer_workload.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    mut backend: Box<dyn DB>,
+    n_accounts: u64,
+    additional_initial_capacity: u64,
+    timer: &mut Timer,
+) {
+    let accounts = (0..n_accounts * 2)
```
Comment (rphmeier) [inline xtask/src/transfer_workload.rs:12]: This creates twice as many accounts as the parameter. The caller might be confused as a result.

Comment (gabriele-0201) [reply xtask/src/transfer_workload.rs:12]: Oh, I misnamed the argument of the function. It should not be `n_accounts` but rather the `size` of the workload. Therefore, considering that a transfer workload involves 2 reads and 2 writes between two accounts (based on what can be done in the database, only perform a vector of actions), I believed we could achieve a transfer workload by transferring balance between `size_transfer_workload * 2` accounts

File: xtask/src/transfer_workload.rs (hunk 2 above)
Comment (gabriele-0201) [reply xtask/src/transfer_workload.rs:6]: Interesting metrics! Initially, I envisioned it differently when trying to find a good common divisor for the custom workload. I am not sure how to handle the percentage of warm/cold storage without making the CLI messy.

I will attempt to further decouple the transfer workload from the custom workload, which may make it easier to develop new workloads and address those three points!

**Review by rphmeier — APPROVED** [review]

**Comment by rphmeier** [comment]
merge when ready

**Pull request merged**